src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -647,11 +648,11 @@ public void run() {
                         numRetries = 0;
                     }
                 } catch (IOException e) {
-                    if (shutdown) {
-                        break;
-                    }
                     LOG.error("Exception while listening", e);
-                    numRetries++;
+                    if (!(e instanceof SocketTimeoutException)) {
+                        numRetries++;
+                    }
+                }finally {
```

Comment:
1. add a space between **}** and **finally**
2. why we need to move some codes about closing **ServerSocket** in **catch** code block to **finally** code block? this will 
   make codes in **Line677-Line685** redundant
3. If **numRetries** don't count **SocketTimeoutExceptions**.is this code facing an endless loop if SocketTimeoutExceptions always happen for a long time? this way of handling SocketTimeoutException is appropriate?